File: apps/android-pos/app/src/main/java/com/pos1/app/data/model/ApiModels.kt
package com.pos1.app.data.model

import kotlinx.serialization.Serializable

/* ═══════════════════════════════════════════════════════════════
   API Response Models — matching backend schema exactly
   ═══════════════════════════════════════════════════════════════ */

// ── Auth ──

@Serializable
data class LoginResponse(
    val accessToken: String,
    val refreshToken: String,
    val user: UserInfo,
)

@Serializable
data class UserInfo(
    val id: String,
    val email: String,
    val displayName: String? = null,
    val roles: List<String> = emptyList(),
    val branchIds: List<String> = emptyList(),
)

// ── Snapshot ──

@Serializable
data class SnapshotData(
    val snapshotAt: String = "",
    val categories: List<CategoryDto> = emptyList(),
    val products: List<ProductDto> = emptyList(),
    val tables: List<TableDto> = emptyList(),
    val settings: BranchSettings = BranchSettings(),
)

@Serializable
data class BranchSettings(
    val taxRateBps: Int = 1500,
    val serviceChargeBps: Int = 0,
    val waiterCallCooldownSec: Int = 30,
    val isQrOrderingEnabled: Boolean = false,
    val isWaiterCallEnabled: Boolean = false,
)

@Serializable
data class CategoryDto(
    val id: String,
    val nameAr: String,
    val nameEn: String? = null,
    val sortOrder: Int = 0,
)

@Serializable
data class ProductDto(
    val id: String,
    val nameAr: String,
    val nameEn: String? = null,
    val basePrice: Double,
    val categoryId: String = "",
    val category: CategoryRef? = null,
    val requiresKitchen: Boolean = true,
    val isActive: Boolean = true,
    val imageUrl: String? = null,
    val modifierGroups: List<ModifierGroupDto> = emptyList(),
)

@Serializable
data class CategoryRef(
    val nameAr: String = "عام",
    val nameEn: String? = null,
)

@Serializable
data class ModifierGroupDto(
    val id: String,
    val nameAr: String,
    val nameEn: String? = null,
    val minSelections: Int = 0,
    val maxSelections: Int = 1,
    val options: List<ModifierOptionDto> = emptyList(),
)

@Serializable
data class ModifierOptionDto(
    val id: String,
    val nameAr: String,
    val nameEn: String? = null,
    val priceAdjustment: Double = 0.0,
    val isDefault: Boolean = false,
)

// ── Tables ──

@Serializable
data class TableDto(
    val id: String,
    val code: String,
    val seats: Int = 4,
    val isActive: Boolean = true,
    val tableAreaId: String? = null,
    val area: TableAreaDto? = null,
)

@Serializable
data class TableAreaDto(
    val id: String,
    val nameAr: String,
    val nameEn: String? = null,
)

// ── Orders ──

@Serializable
data class OrderDto(
    val id: String,
    val orderNo: String,
    val type: String = "TAKEAWAY",
    val status: String = "CONFIRMED",
    val totalAmount: Double = 0.0,
    val discountAmount: Double = 0.0,
    val taxAmount: Double = 0.0,
    val serviceChargeAmount: Double = 0.0,
    val grandTotal: Double = 0.0,
    val tableId: String? = null,
    val table: OrderTableRef? = null,
    val items: List<OrderItemDto> = emptyList(),
    val payments: List<PaymentDto> = emptyList(),
    val createdAt: String = "",
    val updatedAt: String? = null,
)

@Serializable
data class OrderTableRef(
    val id: String = "",
    val code: String = "",
)

@Serializable
data class OrderItemDto(
    val id: String,
    val productId: String,
    val itemNameAr: String = "",
    val qty: Int = 1,
    val unitPrice: Double = 0.0,
    val totalPrice: Double = 0.0,
    val modifiers: List<OrderItemModifierDto> = emptyList(),
)

@Serializable
data class OrderItemModifierDto(
    val id: String = "",
    val nameAr: String = "",
    val priceAdjustment: Double = 0.0,
)

@Serializable
data class PaymentDto(
    val id: String,
    val method: String = "CASH",
    val amount: Double = 0.0,
    val status: String = "CAPTURED",
)

// ── KDS ──

@Serializable
data class KdsTicketDto(
    val id: String,
    val status: String = "NEW",
    val createdAt: String = "",
    val items: List<KdsTicketItemDto> = emptyList(),
    val order: KdsOrderRef? = null,
)

@Serializable
data class KdsTicketItemDto(
    val id: String = "",
    val status: String = "NEW",
    val station: KdsStationRef? = null,
    val orderItem: KdsOrderItemRef? = null,
)

@Serializable
data class KdsStationRef(
    val id: String = "",
    val nameAr: String = "",
)

@Serializable
data class KdsOrderItemRef(
    val id: String = "",
    val itemNameAr: String = "",
    val qty: Int = 1,
)

@Serializable
data class KdsOrderRef(
    val id: String = "",
    val orderNo: String = "",
    val table: OrderTableRef? = null,
)

// ── Waiter ──

@Serializable
data class WaiterCallDto(
    val id: String,
    val reason: String = "",
    val status: String = "CREATED",
    val table: OrderTableRef? = null,
    val createdAt: String = "",
)

@Serializable
data class ReadyOrderDto(
    val id: String,
    val order: KdsOrderRef? = null,
)

// ── Coupon ──

@Serializable
data class CouponValidation(
    val valid: Boolean = false,
    val couponId: String = "",
    val code: String = "",
    val type: String = "PERCENTAGE",
    val discount: Double = 0.0,
)

// ── Loyalty ──

@Serializable
data class LoyaltyAccount(
    val id: String = "",
    val phone: String = "",
    val points: Int = 0,
    val lifetimePoints: Int = 0,
)

File: ios/ios/Data/Models/ApiModels.swift
import Foundation

// MARK: - Auth

struct LoginResponseWrapper: Codable {
    let ok: Bool
    let data: LoginResponse
}

struct LoginResponse: Codable {
    let accessToken: String
    let refreshToken: String
    let user: UserInfo
}

struct UserInfo: Codable {
    let id: String
    let email: String
    let displayName: String?
    let roles: [String]
    let branchIds: [String]
}

// MARK: - Snapshot

struct SnapshotData: Codable {
    var snapshotAt: String = ""
    var categories: [CategoryDto] = []
    var products: [ProductDto] = []
    var tables: [TableDto] = []
    var settings: BranchSettings = BranchSettings()
}

struct BranchSettings: Codable {
    @FlexibleInt var taxRateBps: Int = 1500
    @FlexibleInt var serviceChargeBps: Int = 0
    @FlexibleInt var waiterCallCooldownSec: Int = 30
    var isQrOrderingEnabled: Bool = false
    var isWaiterCallEnabled: Bool = false
}

struct CategoryDto: Codable, Identifiable {
    let id: String
    let nameAr: String
    let nameEn: String?
    @FlexibleInt var sortOrder: Int = 0
}

struct ProductDto: Codable, Identifiable {
    let id: String
    let nameAr: String
    let nameEn: String?
    @FlexibleDouble var basePrice: Double = 0.0
    var categoryId: String = ""
    var category: CategoryRef?
    var requiresKitchen: Bool = true
    var isActive: Bool = true
    let imageUrl: String?
    var modifierGroups: [ModifierGroupDto] = []
}

struct CategoryRef: Codable {
    var nameAr: String = "عام"
    let nameEn: String?
}

struct ModifierGroupDto: Codable, Identifiable {
    let id: String
    let nameAr: String
    let nameEn: String?
    @FlexibleInt var minSelections: Int = 0
    @FlexibleInt var maxSelections: Int = 1
    var options: [ModifierOptionDto] = []
}

struct ModifierOptionDto: Codable, Identifiable {
    let id: String
    let nameAr: String
    let nameEn: String?
    @FlexibleDouble var priceAdjustment: Double = 0.0
    var isDefault: Bool = false
}

// MARK: - Tables

struct TableDto: Codable, Identifiable {
    let id: String
    let code: String
    @FlexibleInt var seats: Int = 4
    var isActive: Bool = true
    let tableAreaId: String?
    let area: TableAreaDto?
}

struct TableAreaDto: Codable, Identifiable {
    let id: String
    let nameAr: String
    let nameEn: String?
}

// MARK: - Orders

struct OrderDto: Decodable, Identifiable {
    let id: String
    let orderNo: String
    var type: String = "TAKEAWAY"
    var status: String = "CONFIRMED"
    @FlexibleDouble var totalAmount: Double = 0.0
    @FlexibleDouble var discountAmount: Double = 0.0
    @FlexibleDouble var taxAmount: Double = 0.0
    @FlexibleDouble var serviceChargeAmount: Double = 0.0
    @FlexibleDouble var grandTotal: Double = 0.0
    let tableId: String?
    let table: OrderTableRef?
    var items: [OrderItemDto] = []
    var payments: [PaymentDto] = []
    var createdAt: String = ""
    let updatedAt: String?

    enum CodingKeys: String, CodingKey {
        case id
        case orderNo
        case type
        case status
        case totalAmount
        case discountAmount
        case taxAmount
        case serviceChargeAmount
        case serviceCharge
        case grandTotal
        case tableId
        case table
        case items
        case payments
        case createdAt
        case updatedAt
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        id = try container.decode(String.self, forKey: .id)
        orderNo = try container.decode(String.self, forKey: .orderNo)
        type = try container.decodeIfPresent(String.self, forKey: .type) ?? "TAKEAWAY"
        status = try container.decodeIfPresent(String.self, forKey: .status) ?? "CONFIRMED"
        _totalAmount = try container.decodeIfPresent(FlexibleDouble.self, forKey: .totalAmount) ?? FlexibleDouble(wrappedValue: 0.0)
        _discountAmount = try container.decodeIfPresent(FlexibleDouble.self, forKey: .discountAmount) ?? FlexibleDouble(wrappedValue: 0.0)
        _taxAmount = try container.decodeIfPresent(FlexibleDouble.self, forKey: .taxAmount) ?? FlexibleDouble(wrappedValue: 0.0)
        _serviceChargeAmount =
            try container.decodeIfPresent(FlexibleDouble.self, forKey: .serviceChargeAmount) ??
            container.decodeIfPresent(FlexibleDouble.self, forKey: .serviceCharge) ??
            FlexibleDouble(wrappedValue: 0.0)
        _grandTotal =
            try container.decodeIfPresent(FlexibleDouble.self, forKey: .grandTotal) ??
            container.decodeIfPresent(FlexibleDouble.self, forKey: .totalAmount) ??
            FlexibleDouble(wrappedValue: 0.0)
        tableId = try container.decodeIfPresent(String.self, forKey: .tableId)
        table = try container.decodeIfPresent(OrderTableRef.self, forKey: .table)
        items = try container.decodeIfPresent([OrderItemDto].self, forKey: .items) ?? []
        payments = try container.decodeIfPresent([PaymentDto].self, forKey: .payments) ?? []
        createdAt = try container.decodeIfPresent(String.self, forKey: .createdAt) ?? ""
        updatedAt = try container.decodeIfPresent(String.self, forKey: .updatedAt)
    }
}

struct OrderTableRef: Codable {
    var id: String = ""
    var code: String = ""
}

struct OrderItemDto: Decodable, Identifiable {
    let id: String
    let productId: String
    var itemNameAr: String = ""
    @FlexibleInt var qty: Int = 1
    @FlexibleDouble var unitPrice: Double = 0.0
    @FlexibleDouble var totalPrice: Double = 0.0
    var modifiers: [OrderItemModifierDto] = []

    enum CodingKeys: String, CodingKey {
        case id
        case productId
        case itemNameAr
        case qty
        case unitPrice
        case totalPrice
        case lineTotal
        case modifiers
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        id = try container.decode(String.self, forKey: .id)
        productId = try container.decode(String.self, forKey: .productId)
        itemNameAr = try container.decodeIfPresent(String.self, forKey: .itemNameAr) ?? ""
        _qty = try container.decodeIfPresent(FlexibleInt.self, forKey: .qty) ?? FlexibleInt(wrappedValue: 1)
        _unitPrice = try container.decodeIfPresent(FlexibleDouble.self, forKey: .unitPrice) ?? FlexibleDouble(wrappedValue: 0.0)
        _totalPrice =
            try container.decodeIfPresent(FlexibleDouble.self, forKey: .totalPrice) ??
            container.decodeIfPresent(FlexibleDouble.self, forKey: .lineTotal) ??
            FlexibleDouble(wrappedValue: 0.0)
        modifiers = try container.decodeIfPresent([OrderItemModifierDto].self, forKey: .modifiers) ?? []
    }
}

struct OrderItemModifierDto: Codable, Identifiable {
    var id: String = ""
    var nameAr: String = ""
    @FlexibleDouble var priceAdjustment: Double = 0.0
}

struct PaymentDto: Codable, Identifiable {
    let id: String
    var method: String = "CASH"
    @FlexibleDouble var amount: Double = 0.0
    var status: String = "CAPTURED"
}

// MARK: - KDS

struct KdsTicketDto: Codable, Identifiable {
    let id: String
    var status: String = "NEW"
    var createdAt: String = ""
    var items: [KdsTicketItemDto] = []
    let order: KdsOrderRef?
}

struct KdsTicketItemDto: Codable, Identifiable {
    var id: String = ""
    var status: String = "NEW"
    let station: KdsStationRef?
    let orderItem: KdsOrderItemRef?
}

struct KdsStationRef: Codable, Identifiable {
    var id: String = ""
    var nameAr: String = ""
}

struct KdsOrderItemRef: Codable, Identifiable {
    var id: String = ""
    var itemNameAr: String = ""
    @FlexibleInt var qty: Int = 1
}

struct KdsOrderRef: Codable, Identifiable {
    var id: String = ""
    // Add other fields if necessary based on usage
}
